scripts/policy/frameworks/cluster/agent/main.zeek

```diff
@@ -22,6 +22,12 @@ global g_nodes: table[string] of ClusterController::Types::Node;
 # new configurations.
 global g_data_cluster: table[string] of Supervisor::ClusterEndpoint;
 
+# Whether we currenty keep notifying the controller that we're here.
+# We stop once the controller responds back, and resume when we've
+# lost the peering.
```

Comment:
I can't quite align the comment with the code: "keep notifying" suggest to me that we continue to send the even unsolicited, but if I'm reading the code right, we send it only on demand after receiving the hello from cluster.

Reply:
It should work as follows:
- Regardless of the direction of connection establishment, the `Broker::peer_added` event fires upon connectivity, and the agent starts sending the `notify_agent_hello` events. (There's no point in sending them before there's connectivity.)
- These pings continue until the controller sends `notify_controller_hello`, which is the signal to the agent that it's now welcome, as per the currently active configuration. The agent stops pinging, and will resume upon the next `peer_added`.

But! Looking at this now, seems we can simplify: the point of `notify_agent_hello` with the current approach is really to convey to the controller that an agent is ready. There's no need for this ongoing pinging: we could have a request/response event pair `agent_welcome_request` and `agent_welcome_response` ... the controller sends the request to agents it wants to see in service, and tracks the response event like it now tracks `notify_agent_hello`. And for agents the controller no longer requires it can send a `agent_farewell`, no response needed, which'll cause the agent to stop its cluster nodes.

What do you think? I'll give this a go now.

Reply:
Yep, that simplification sounds good (and the description makes sense, I must have gotten confused with event semantics, per the other comment :-)

Reply:
I've now worked this in. I simply put the change in an additional commit since I got afraid that it's too painful to sort through each subsequent commit if I rebase. `notify_agent_hello` is now really just a checkin to convey after a peering that we're dealing with an agent, and the controller uses `agent_welcome` request/responses to notify an agent that it's now going to participate in the running cluster, while a `agent_standby` request/response tells it to free it's cluster state and wait for further instructions.

There's one more minor change in this: I'm no longer trying to `unpeer()` unused agents, i.e., an agent can simply stick around even if it's not currently useful. This was imperfect anyway since one can only do this if one also previously `peer()`'ed, and it seems that when managing agents this way, there'll often be a need to restart it anyway.

I remembered that Benjamin has a nice way of visualizing protocol interactions via `plantuml` ... I'll see if I can get that to work so I can use it in comments.

Reply:
Looks good!